Specify when clients can serve block and sidecars in byRoot RPC methods (#3551)
ethereum/consensus-specs: specs/phase0/p2p-interface.md

Clients MUST respond with at least one block, if they have it.
Clients MAY limit the number of blocks in the response.

> Clients CAN consider including a block in the response as soon as it it passes the gossip validation rules.

Reviewer: ```suggestion
Clients MAY consider including a block in the response as soon as it it passes the gossip validation rules.
```

> Clients MAY consider including a block in the response as soon as it it passes the gossip validation rules.

Reviewer: ```suggestion
Clients MAY consider including a block in the response as soon as it passes the gossip validation rules.
```

Clients MUST NOT respond with blocks that fail gossip validation rules.

> Clients SHOULD NOT include blocks that fail `state_transition(block)`.

Reviewer: Nit: `state_transition` has more arguments than just a block.
(this is probably my fault as I have used it this way in one of my comments)
```suggestion
Clients SHOULD NOT respond with blocks that fail beacon chain state transition.
```

> Clients MUST NOT respond with blocks that failed `fork_choice.on_block`.

Author: I have mixed feeling on this.
I remember @djrtwo mentioning to avoid serving blocks that turns out to be invalid, but I might recall incorrectly.

An argument in favour is that it seems odd to have clients serving blocks that turned out to be invalid for them.
As the time pass by, the client can from better ideas on a particular block, so makes sense to me that most recent byRoot responses leverage these additional evaluations applied after initial gossip rules.

> Clients MUST NOT respond with blocks that failed `fork_choice.on_block`.

Reviewer 2: yeah, I don't think it's a MUST NOT. I think it's a MAY NOT ( you can safely prune these) but I'm not 100% sure if it needs to be called out.

Also, I'm a fan of the state transition being hte filter instead of the on_block as on_block has many timing and stateful conditions here

> Clients MUST NOT respond with blocks that failed `fork_choice.on_block`.

Author: > Also, I'm a fan of the state transition being hte filter instead of the on_block as on_block has many timing and stateful conditions here

I see your point. I only see a potential issue due to the fact that in `on_block` as written in spec, `is_data_available`  comes before the state transition, and if we want clients keep serving blocks with unavailable blobs, seems like they could do that even without checking the state transition. But clients implementations are most likely not following that order, so let's change..

>  I don't think it's a MUST NOT. I think it's a MAY NOT
agreed.

Clients MUST respond with at least one block, if they have it.
Clients MAY limit the number of blocks in the response.

Clients MAY include a block in the response as soon as it passes the gossip validation rules.

> Clients SHOULD NOT respond with blocks that fail beacon chain state transition.

Reviewer 3: sounds good, it would have been good to enumerate the rules so as to be super clear, but nevertheless lgtm

> Clients SHOULD NOT respond with blocks that fail beacon chain state transition.

Reviewer 2: ```suggestion
Clients SHOULD NOT respond with blocks that fail the beacon chain state transition.
```
